File: BackEnd/src/domain/entities/Plant.ts
/**
 * @file Plant.ts
 * @description Entidad de dominio que representa una planta del usuario.
 * Contiene lógica de negocio pura para cálculo de próximo riego y ajustes de override.
 * @module Plants
 * @layer Domain
 */

/**
 * Ubicación física de la planta.
 */
export type PlantLocation = 'Interior' | 'Exterior';

/**
 * Necesidad de luz de la planta.
 */
export type LightNeed = 'Low' | 'Medium' | 'High';

/**
 * Estado de un reset de nextWatering provocado por una previsión de lluvia.
 * Se rellena cuando _processWeather aplica un reset por previsión, y se
 * consume al día siguiente en _processYesterdayRain:
 *   - Si la lluvia se confirmó (history >= minRainfallMm) → limpiar y notificar.
 *   - Si NO se confirmó → restaurar previousNextWatering, limpiar y notificar.
 */
export interface PendingRainAdjustment {
  /** Fecha en la que se aplicó el reset por previsión */
  resetAt: Date;
  /** Valor de `nextWatering` ANTES del reset (null si la planta no tenía) */
  previousNextWatering: Date | null;
  /** Probabilidad/precipitación esperada que motivó el reset (informativo) */
  expectedMm: number;
  /** Coordenadas usadas para consultar history al día siguiente */
  locationLat: number;
  locationLon: number;
}

/**
 * Override puntual para alterar la frecuencia de riego en un rango de fechas.
 * Permite que condiciones climáticas o del usuario ajusten el comportamiento sin
 * modificar la configuración base de la planta.
 */
export interface PlantOverride {
  /** Fecha de inicio del override (inclusive) */
  fromDate: Date;
  /** Fecha de fin del override (inclusive) */
  toDate: Date;
  /** Frecuencia de riego ajustada en días */
  wateringFrequencyDays: number;
  /** Motivo del override (ej: "Período de lluvias", "Vacaciones") */
  reason?: string;
}

/**
 * Entidad de dominio Plant.
 * Representa el estado válido de una planta dentro del sistema.
 * Los mappers convierten entre esta entidad y los documentos de MongoDB.
 */
export class Plant {
  /** Identificador único (ObjectId de MongoDB serializado como string) */
  readonly id: string;

  /** Id del usuario propietario */
  readonly userId: string;

  /** Nombre personalizado de la planta */
  readonly name: string;

  /** Id de la especie asociada (opcional) */
  readonly speciesId?: string;

  /** URL de la foto en Cloudinary */
  readonly photo?: string;

  /** Ubicación de la planta */
  readonly location: PlantLocation;

  /** Ciudad/municipio donde se encuentra la planta (catálogo SPAIN_LOCATIONS) */
  readonly plantLocation?: string;

  /** Latitud geográfica de la ubicación de la planta */
  readonly plantLocationLat?: number;

  /** Longitud geográfica de la ubicación de la planta */
  readonly plantLocationLon?: number;

  /** Frecuencia base de riego en días */
  readonly wateringFrequency: number;

  /** Necesidad de luz */
  readonly lightNeed: LightNeed;

  /** Frecuencia de poda en días (opcional) */
  readonly pruningFrequency?: number;

  /** Notas libres del usuario */
  readonly notes?: string;

  /** Próxima fecha de riego calculada */
  readonly nextWatering?: Date;

  /** Próxima fecha de poda calculada */
  readonly nextPruning?: Date;

  /** Si true, el cálculo de riego considera datos meteorológicos */
  readonly considerWeatherForWatering: boolean;

  /** Overrides puntuales de frecuencia de riego */
  readonly overrides?: PlantOverride[];

  /** Reset de nextWatering pendiente de confirmación por history */
  readonly pendingRainAdjustment?: PendingRainAdjustment;

  /** Fecha de creación */
  readonly createdAt: Date;

  /** Fecha de última modificación */
  readonly updatedAt: Date;

  /** Fecha de borrado lógico. Si está definida, la planta está eliminada */
  readonly deletedAt?: Date | null;

  constructor(params: {
    id: string;
    userId: string;
    name: string;
    speciesId?: string;
    photo?: string;
    location: PlantLocation;
    plantLocation?: string;
    plantLocationLat?: number;
    plantLocationLon?: number;
    wateringFrequency: number;
    lightNeed: LightNeed;
    pruningFrequency?: number;
    notes?: string;
    nextWatering?: Date;
    nextPruning?: Date;
    considerWeatherForWatering: boolean;
    overrides?: PlantOverride[];
    pendingRainAdjustment?: PendingRainAdjustment;
    createdAt: Date;
    updatedAt: Date;
    deletedAt?: Date | null;
  }) {
    this.id = params.id;
    this.userId = params.userId;
    this.name = params.name;
    this.speciesId = params.speciesId;
    this.photo            = params.photo;
    this.location         = params.location;
    this.plantLocation    = params.plantLocation;
    this.plantLocationLat = params.plantLocationLat;
    this.plantLocationLon = params.plantLocationLon;
    this.wateringFrequency = params.wateringFrequency;
    this.lightNeed = params.lightNeed;
    this.pruningFrequency = params.pruningFrequency;
    this.notes = params.notes;
    this.nextWatering = params.nextWatering;
    this.nextPruning = params.nextPruning;
    this.considerWeatherForWatering = params.considerWeatherForWatering;
    this.overrides = params.overrides;
    this.pendingRainAdjustment = params.pendingRainAdjustment;
    this.createdAt = params.createdAt;
    this.updatedAt = params.updatedAt;
    this.deletedAt = params.deletedAt;
  }

  /**
   * Calcula la próxima fecha de riego a partir de una fecha de referencia.
   * Si hay un override activo en esa fecha, usa su frecuencia ajustada.
   *
   * @param now — Fecha de referencia (por defecto: ahora).
   * @returns Fecha del próximo riego.
   */
  calculateNextWatering(now: Date = new Date()): Date {
    const activeOverride = this.overrides?.find(
      (o) => o.fromDate <= now && o.toDate >= now,
    );

    const frequencyDays = activeOverride
      ? activeOverride.wateringFrequencyDays
      : this.wateringFrequency;

    const next = new Date(now);
    next.setDate(next.getDate() + frequencyDays);
    return next;
  }

  /**
   * Calcula la próxima fecha de poda a partir de una fecha de referencia.
   * Devuelve undefined si la planta no tiene frecuencia de poda configurada.
   *
   * @param now — Fecha de referencia (por defecto: ahora).
   * @returns Fecha del próximo poda o undefined.
   */
  calculateNextPruning(now: Date = new Date()): Date | undefined {
    if (!this.pruningFrequency) return undefined;
    const next = new Date(now);
    next.setDate(next.getDate() + this.pruningFrequency);
    return next;
  }

  /**
   * Determina si hay un override activo en una fecha dada.
   * Útil para que el ReminderService ajuste la frecuencia antes de enviar recordatorios.
   *
   * @param date — Fecha a comprobar.
   * @returns El override activo o undefined.
   */
  getActiveOverride(date: Date = new Date()): PlantOverride | undefined {
    return this.overrides?.find((o) => o.fromDate <= date && o.toDate >= date);
  }

  /**
   * Indica si la planta necesita riego hoy (nextWatering <= now).
   */
  get needsWateringToday(): boolean {
    if (!this.nextWatering) return false;
    return this.nextWatering <= new Date();
  }
}

File: FrontEnd/lib/domain/entities/plant.dart
/// @file plant.dart
/// @description Entidad de dominio Plant. Representa una planta del usuario.
/// Objeto puro de Dart, sin dependencias de Flutter ni paquetes externos.
/// Los mappers convierten PlantModel ↔ Plant.
/// @module Plants
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// PLANT ENTITY
// ═══════════════════════════════════════════════════════════════════════════════

/// Entidad de dominio que representa una planta del usuario.
///
/// Todos los campos son inmutables (final). Usar [copyWith] para actualizaciones.
/// Sin lógica de serialización — eso pertenece a PlantModel (data layer).
class Plant {
  const Plant({
    required this.id,
    required this.userId,
    required this.name,
    required this.wateringFrequencyDays,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.speciesId,
    this.photo,
    this.location,
    this.plantLocation,
    this.plantLocationLat,
    this.plantLocationLon,
    this.notes,
    this.lightNeed,
    this.pruningFrequency,
    this.nextPruning,
    this.considerWeatherForWatering = false,
    this.lastWatered,
    this.nextWatering,
    this.careOverrides,
  });

  /// Identificador único de la planta (MongoDB ObjectId como String).
  final String id;

  /// ID del usuario propietario.
  final String userId;

  /// Nombre visible de la planta.
  final String name;

  /// ID de la especie asociada, o null si no tiene especie asignada.
  final String? speciesId;

  /// URL de la foto (Cloudinary), o null si no tiene.
  final String? photo;

  /// Ubicación física de la planta (p.ej. "Balcón sur"), o null.
  final String? location;

  /// Ciudad/municipio donde se encuentra la planta (del catálogo de capitales), o null.
  final String? plantLocation;

  /// Latitud geográfica de la ubicación de la planta, o null.
  final double? plantLocationLat;

  /// Longitud geográfica de la ubicación de la planta, o null.
  final double? plantLocationLon;

  /// Notas libres del usuario, o null.
  final String? notes;

  /// Frecuencia de riego en días (p.ej. 7 = semanal).
  final int wateringFrequencyDays;

  /// Necesidad de luz de la planta ('Low', 'Medium', 'High'), o null.
  final String? lightNeed;

  /// Frecuencia de poda en días, o null si no aplica.
  final int? pruningFrequency;

  /// Fecha UTC de la próxima poda calculada, o null si no aplica.
  final DateTime? nextPruning;

  /// true si el riego debe considerar el pronóstico del tiempo.
  final bool considerWeatherForWatering;

  /// Fecha UTC del último riego registrado, o null si nunca se ha regado.
  final DateTime? lastWatered;

  /// Fecha UTC del próximo riego calculado, o null si no aplica.
  final DateTime? nextWatering;

  /// true si la planta está activa (no eliminada lógicamente).
  final bool isActive;

  /// Anulaciones manuales de los parámetros de cuidado de la especie, o null.
  final PlantCareOverrides? careOverrides;

  /// Fecha de creación del registro.
  final DateTime createdAt;

  /// Fecha de última actualización del registro.
  final DateTime updatedAt;

  // ─── Helpers ─────────────────────────────────────────────────────────────────

  /// true si la planta necesita riego hoy o ya estaba atrasada.
  bool get needsWatering {
    if (nextWatering == null) return false;
    final today = DateTime.utc(
      DateTime.now().year,
      DateTime.now().month,
      DateTime.now().day,
    );
    return !nextWatering!.isAfter(today);
  }

  /// Días restantes hasta el próximo riego. Negativo = ya debería haberse regado.
  int get daysUntilNextWatering {
    if (nextWatering == null) return 0;
    final today = DateTime.utc(
      DateTime.now().year,
      DateTime.now().month,
      DateTime.now().day,
    );
    return nextWatering!.difference(today).inDays;
  }

  /// true si la planta tiene una especie asignada.
  bool get hasSpecies => speciesId != null && speciesId!.isNotEmpty;

  /// true si la planta tiene foto.
  bool get hasPhoto => photo != null && photo!.isNotEmpty;

  // ─── copyWith ────────────────────────────────────────────────────────────────

  /// Devuelve una copia de la planta con los campos indicados modificados.
  Plant copyWith({
    String?             id,
    String?             userId,
    String?             name,
    String?             speciesId,
    String?             photo,
    String?             location,
    String?             plantLocation,
    double?             plantLocationLat,
    double?             plantLocationLon,
    String?             notes,
    int?                wateringFrequencyDays,
    String?             lightNeed,
    int?                pruningFrequency,
    DateTime?           nextPruning,
    bool?               considerWeatherForWatering,
    DateTime?           lastWatered,
    DateTime?           nextWatering,
    bool?               isActive,
    PlantCareOverrides? careOverrides,
    DateTime?           createdAt,
    DateTime?           updatedAt,
  }) {
    return Plant(
      id:                    id                    ?? this.id,
      userId:                userId                ?? this.userId,
      name:                  name                  ?? this.name,
      speciesId:             speciesId             ?? this.speciesId,
      photo:                 photo                 ?? this.photo,
      location:              location              ?? this.location,
      plantLocation:         plantLocation         ?? this.plantLocation,
      plantLocationLat:      plantLocationLat      ?? this.plantLocationLat,
      plantLocationLon:      plantLocationLon      ?? this.plantLocationLon,
      notes:                 notes                 ?? this.notes,
      wateringFrequencyDays: wateringFrequencyDays ?? this.wateringFrequencyDays,
      lightNeed:             lightNeed             ?? this.lightNeed,
      pruningFrequency:      pruningFrequency      ?? this.pruningFrequency,
      nextPruning:           nextPruning           ?? this.nextPruning,
      considerWeatherForWatering: considerWeatherForWatering ?? this.considerWeatherForWatering,
      lastWatered:           lastWatered           ?? this.lastWatered,
      nextWatering:          nextWatering          ?? this.nextWatering,
      isActive:              isActive              ?? this.isActive,
      careOverrides:         careOverrides         ?? this.careOverrides,
      createdAt:             createdAt             ?? this.createdAt,
      updatedAt:             updatedAt             ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Plant && other.id == id);

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'Plant(id: $id, name: $name, userId: $userId)';
}

// ═══════════════════════════════════════════════════════════════════════════════
// PLANT CARE OVERRIDES
// ═══════════════════════════════════════════════════════════════════════════════

/// Anulación manual de los parámetros de cuidado heredados de la especie.
/// El usuario puede ajustar la frecuencia de riego o la necesidad de luz.
class PlantCareOverrides {
  const PlantCareOverrides({
    this.wateringFrequencyDays,
    this.lightNeed,
  });

  /// Frecuencia de riego personalizada en días, o null si usa la de la especie.
  final int?    wateringFrequencyDays;

  /// Necesidad de luz personalizada ('Low', 'Medium', 'High'), o null.
  final String? lightNeed;

  PlantCareOverrides copyWith({
    int?    wateringFrequencyDays,
    String? lightNeed,
  }) {
    return PlantCareOverrides(
      wateringFrequencyDays: wateringFrequencyDays ?? this.wateringFrequencyDays,
      lightNeed:             lightNeed             ?? this.lightNeed,
    );
  }
}
